File: srcs/django-files/backend/tpong/static/tpong/Paddle.js
import { Util } from "./Util.js";
import { vec2 } from "./vec2.js";
import { Color } from "./Color.js";
import InputManager from "./InputManager.js";
import collisionManager from "./CollisionManager.js";

export class Paddle
{
	net =	{
		CONTROLLING:'controlling',
		REPLICATED:'replicated'
	}
	
	constructor(side, size)
	{
		this.type = "paddle"
		this.side = side;
		this.sizeNormal = size;
		this.heightNormal = 0.5;
		this.serverHeightNormal = 0.5;
		this.velocityNormal = 0.0;
		this.maxVelocity = 0.7
		this.cuid = collisionManager.addCollidableRect(this.getWindowNormal(), this.sizeNormal, false, this.onCollision.bind(this), this)
		this.networkingState = this.net.CONTROLLING;
		this.color = this.side == "left" ? new Color(15, 125, 252) : new Color(255, 42, 42)
	}

	getWindowNormal()
	{
		var x = 0;
		if (this.side == 'right')
			x = 1 - this.sizeNormal.x;
		return new vec2(x, this.heightNormal * (1 - this.sizeNormal.y))
	}

	syncDOM()
	{
		var paddle = Util.getPaddleHtmlObject(this.side);
		const windowBoundingBox = Util.getGameWindowHtmlObject().getBoundingClientRect();
		var ballBoundingBox = Util.getBallHtmlObject().getBoundingClientRect();
		var pos = this.getWindowNormal();
		var aspectRatioW = windowBoundingBox.width / windowBoundingBox.height

		var aspectFix = (ballBoundingBox.height * aspectRatioW) - ballBoundingBox.height

		pos = pos.multiply(new vec2(windowBoundingBox.width, windowBoundingBox.height ));
		// pos = pos.add(new vec2(windowBoundingBox.left + document.documentElement.scrollLeft, windowBoundingBox.top + document.documentElement.scrollTop));
		paddle.style.backgroundColor = this.color.toRgbCss()

		paddle.style.width = this.sizeNormal.x * windowBoundingBox.width  + 'px';
		paddle.style.height = this.sizeNormal.y * windowBoundingBox.height + 'px';
		paddle.style.position = 'abolute';
		paddle.style.left = pos.x + 'px';	
		paddle.style.top =  pos.y +'px';
	}

	simulate(delta)
	{
		//if (this.networkingState == this.net.CONTROLLING)
		//	this.processInput();
		this.serverHeightNormal = this.serverHeightNormal + this.velocityNormal * delta;
		this.serverHeightNormal = Math.max(Math.min(this.serverHeightNormal, 1), 0);
		this.heightNormal = Util.lerp(this.heightNormal, this.serverHeightNormal, delta * 10)
		this.heightNormal = Math.max(Math.min(this.heightNormal, 1), 0);
		collisionManager.updateCollidableObject(this.cuid, this.getWindowNormal());
	}

	reset()
	{
		this.heightNormal = 0.5;
		this.velocityNormal = 0.0;
	}

	onCollision(collision, obj)
	{
		if (!obj)
			return
		if (obj.type == "ball")
		{
			obj.color = this.color.copy()
		}
	}

	processInput()
	{
		var velocityNormal = 0.0;
		if (InputManager.keysDown[this.side == 'right' ? 'ArrowUp' : 'KeyW'])
			velocityNormal -= this.maxVelocity;
		if (InputManager.keysDown[this.side == 'right' ? 'ArrowDown' : 'KeyS'])
			velocityNormal += this.maxVelocity;
		if (velocityNormal == 0.0)
			this.velocityNormal *= 0.8;
		else
			this.velocityNormal += (velocityNormal - this.velocityNormal) * 0.2
		//this.velocityNormal = velocityNormal;
		return velocityNormal != 0
	}

	serialize()
	{
		return {
			objName: 'paddle',
			paddleSide: this.side,
			paddleHeight: this.heightNormal,
			paddleVelocity: this.velocityNormal
		}
	}
}

File: srcs/django-files/backend/tpong/game/engine/gameObjects/Paddle.py
from tpong.game.engine.Vec2 import Vec2
from .Collidable import Collidable
import logging

logger = logging.getLogger(__name__)

class Paddle(Collidable):
	def __init__(self, side) -> None:
		super().__init__('rect')
		self.size = Vec2(0.03, 0.2)
		self.heightNormal = 0.5
		self.side = side
		self.maxVelocity = 0.7
		self.keystate = {
			"up": False,
			"down": False,
		}

	def simulate(self, delta):
		self.radiusNormal = 0.01
		self.heightNormal += self.velocity.y * delta
		self.heightNormal = max(min(self.heightNormal, 1), 0)
		self.updatePosition()
		self.stateChanged = abs(self.velocity.y) > 0.01 

	def updatePosition(self):
		x = 0
		if self.side == 'right':
			x = 1 - self.size.x
		self.position.x = x
		self.position.y = self.heightNormal * (1 - self.size.y)

	def replicate(self, replicationInfo):
		self.heightNormal = replicationInfo['paddleHeight']
		self.velocity.y = replicationInfo['paddleVelocity']

	def serializeState(self):
		return {
			'objName': 'paddle',
			'paddleHeight': self.heightNormal,
			'paddleVelocity': self.velocity.y,
			'paddleSide': self.side,
			'paddleSize': {
				'x': self.size.x,
				'y': self.size.y
			}
		}
	
	def onCollision(self, collision):
		pass

	def processInput(self):
		newVel = 0
		if self.keystate["up"]:
			newVel -= self.maxVelocity
		if self.keystate["down"]:
			newVel += self.maxVelocity
		self.velocity.y = newVel
		self.stateChanged = True
		
	def processRemoteInput(self, remoteInput):
		self.keystate[remoteInput['dir']] = remoteInput['keyDown']
		self.processInput()


	def reset(self):
		self.heightNormal = 0.5
		self.velocity.y = 0